File: connector/php/WookieConnectorService.php
<?php
/*
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 * limitations under the License.
 */

require("WookieConnectorExceptions.php");
require("WookieServerConnection.php");
require("WidgetInstances.php");
require("Widget.php");
require("WidgetInstance.php");
require("User.php");
require("HTTP_Response.php");


class WookieConnectorService {
  private $conn;
  public  $WidgetInstances;
  private $user;
  
  function __construct($url, $apiKey, $sharedDataKey, $loginName, $screenName = null) {
    $this->setConnection(new WookieServerConnection($url, $apiKey, $sharedDataKey));
	$this->setWidgetInstancesHolder();
	$this->setUser($loginName, $screenName);
  }
  
  public function setConnection($newConn) {
    $this->conn = $newConn;
  }
  
  public function getConnection() {
    return $this->conn;
  }
  
  public function setWidgetInstancesHolder() {
	$this->WidgetInstances = new WidgetInstances();
  }
  
  public function setUser($loginName, $screenName = null) {
	if($screenName == null) {
		$screenName = $loginName;
	}
	$this->user = new User($loginName, $screenName);
  }
  
  public function getUser() {
	return $this->user;
  }
  
  /* Do POST request
  /* @return new HTTP_Response instance */
  
  private function do_post_request($url, $data, $optional_headers = null)
  {
	if(is_array($data)) {
		 // convert variables array to string:
		$_data = array();    
		while(list($n,$v) = each($data)){
			$_data[] = urlencode($n)."=".urlencode($v);
		}    
		$data = implode('&', $_data);
	}
	
     $params = array('http' => array(
                  'method' => 'POST',
                  'content' => $data
               ));
	
     if ($optional_headers !== null) {
        $params['http']['header'] = $optional_headers;
     }
     $ctx = @stream_context_create($params);
    // $fp = @@file_get_contents($url, 'rb', false, $ctx);
     $response = @file_get_contents($url, false, $ctx);

     return new HTTP_Response($response, $http_response_header);
  }
  
  
  /**
   * Get or create an instance of a widget.
   * 
   * @param widget
   * @return the ID of the widget instance
   * @throws IOException
   * @throws SimalRepositoryException
   */

  public function getOrCreateInstance($Widget_or_GUID = null) {
    try {
		if(is_object($Widget_or_GUID)) {
			$guid = $Widget_or_GUID->getIdentifier();
		} else {
			$guid = $Widget_or_GUID;
		}
		if($guid == '') {
			throw new WookieConnectorException("No GUID nor Widget object");
		}
		$requestUrl = $this->getConnection()->getURL().'/widgetinstances';
		$request.= '&api_key='.$this->getConnection()->getApiKey();
		$request.= '&servicetype=';
		$request.= '&userid='.$this->getUser()->getLoginName();
		$request.= '&shareddatakey='.$this->getConnection()->getSharedDataKey();
		$request.= '&widgetid='.$guid;
		
		if(!$this->checkURL($requestUrl)) {
			throw new WookieConnectorException("URL for supplied Wookie Server is malformed: ".$requestUrl);
		}	
		$response = $this->do_post_request($requestUrl, $request);
		
		//if instance was created, perform second request to get widget instance
		if($response->getStatusCode() == 201) {
			$response = $this->do_post_request($requestUrl, $request);
		}
		$instance = $this->parseInstance($guid, $response->getResponseText());
		$this->WidgetInstances->put($instance);
		return $instance;
	} catch (WookieConnectorException $e) {
		echo $e->errorMessage();
	}
		return false;
  }
  
  
    /**
   * Record an instance of the given widget.
   * 
   * @param xml description of the instance as returned by the widget server when the widget was instantiated.
   * @return new Widget instance
   */
  private function parseInstance($widgetGuid, $xml) {
	$xmlWidgetData = @simplexml_load_string($xml);
	if(is_object($xmlWidgetData)) {
		$url = (string) $xmlWidgetData->url;
		$title = (string) $xmlWidgetData->title;
		$height = (string) $xmlWidgetData->height;
		$width = (string) $xmlWidgetData->width;
		$maximize = (string) $xmlWidgetData->maximize;
		$instance = new WidgetInstance($url, $widgetGuid, $title, $height, $width, $maximize);
		return $instance;
	} 
	return false;
  }
  
   /**
   * Check if URL is parsable.
   * 
   * @param url
   * @return boolean
   */
  
  private function checkURL($url) {
	$UrlCheck = parse_url($url);
	if($UrlCheck['scheme'] != 'http' || $UrlCheck['host'] == null || $UrlCheck['path'] == null) {
		return false;
	}
	return true;
  }
  
   /**
   * @refactor At time of writing the REST API for adding a participant is broken so we are
   * using the non-REST approach. The code for REST API is commented out and should be used
   * in the future.
   */
   
  public function addParticipant($widgetInstance, $User)  {
	$Url = $this->getConnection()->getURL().'/participants';
	
	try {
		if(!is_object($widgetInstance)) throw new WookieWidgetInstanceException('No Widget instance');
		if(!is_object($User)) throw new WookieConnectorException('No User object');
		
		$data = array(
				'api_key' => $this->getConnection()->getApiKey(),
				'shareddatakey' => $this->getConnection()->getSharedDataKey(),
				'userid' => $this->getUser()->getLoginName(),
				'widgetid' => $widgetInstance->getIdentifier(),
				'participant_id' => $this->getUser()->getLoginName(),
				'participant_display_name' => $User->getScreenName(),
				'participant_thumbnail_url' => $User->getThumbnailUrl(),
		);	
		
		if(!$this->checkURL($Url)) {
			throw new WookieConnectorException("Participants rest URL is incorrect: ".$Url);
		}	
		
		$response = $this->do_post_request($Url, $data);
		$statusCode = $response->getStatusCode();
		
		switch($statusCode) {
		  case 200: //participant already exists
			return true;
			break;
		  case 201:
			return true; //new participant added
			break;
		  case ($statusCode > 201):
			throw new WookieConnectorException($response->headerToString().'<br />'.$response->getResponseText());
			break;
		  default:
			return false;
		}

	} catch (WookieConnectorException $e) {
		echo $e->errorMessage();
	} catch (WookieWidgetInstanceException $e) {
		echo '<b>function.addParticipant:</b> '.$e->getMessage().'<br />';
	}
    
  }
  
   /**
   * Get the array of users for a widget instance
   * @param instance
   * @return an array of users
   * @throws WookieConnectorException
   */
  public function getUsers($widgetInstance) {
	$Url = $this->getConnection()->getURL().'/participants';
	$Users = array();
		try {
			if(!is_object($widgetInstance)) throw new WookieWidgetInstanceException('No Widget instance');
			$request = '?api_key='.$this->getConnection()->getApiKey();
			$request .= '&shareddatakey='.$this->getConnection()->getSharedDataKey();
			$request .= '&userid='.$this->getUser()->getLoginName();
			$request .= '&widgetid='.$widgetInstance->getIdentifier();
			
			if(!$this->checkURL($Url)) {
				throw new WookieConnectorException("Participants rest URL is incorrect: ".$Url);
			}
			
			$response = new HTTP_Response(@file_get_contents($Url.$request), $http_response_header);
			if($response->getStatusCode() > 200) throw new WookieConnectorException($response->headerToString().'<br />'.$response->getResponseText());
			
			$xmlObj = @simplexml_load_string($response->getResponseText());
			
			if(is_object($xmlObj)) {
				foreach($xmlObj->children() as $participant) {
				  $participantAttr = $participant->attributes();
				  
				  $id = (string) $participantAttr->id;
				  $name = (string) $participantAttr->display_name;
				  $thumbnail_url = (string) $participantAttr->thumbnail_url;
				  
				  $newUser = new User($id, $name, $thumbnail_url);
				  array_push($Users, $newUser);
				}
			} else {
				throw new WookieConnectorException('Problem getting participants');
			}
			
			return $Users;
		} catch (WookieWidgetInstanceException $e) {
			echo '<b>function.getUsers:</b> '.$e->getMessage().'<br />';
		} catch (WookieConnectorException $e) {
			echo $e->errorMessage();
		}
  }

 

  /**
   * Get a set of all the available widgets in the server. If there is an error
   * communicating with the server return an empty set, or the set received so
   * far in order to allow the application to proceed. The application should
   * display an appropriate message in this case.
   * 
   * @return array of available widgets
   * @throws WookieConnectorException
   */
   
  public function getAvailableWidgets() {
	$widgets = array();
	try {
		$request = $this->getConnection()->getURL().'/widgets?all=true';
		
		if(!$this->checkURL($request)) {
			throw new WookieConnectorException("URL for Wookie is malformed");
		}
		
		$response = new HTTP_Response(@file_get_contents($request), $http_response_header);
		$xmlObj = @simplexml_load_string($response->getResponseText());
		
		if(is_object($xmlObj)) {
			foreach($xmlObj->children() as $widget) {
				 $id = (string) $widget->attributes()->identifier;
				 $title = (string) $widget->title;
				 $description = (string) $widget->description;
				 $iconURL = (string) $widget->attributes()->icon;
				 if($iconURL == '') {
						$iconURL = (string) 'http://www.oss-watch.ac.uk/images/logo2.gif';
					}
				$Widget = new Widget($id, $title, $description, $iconURL);
				$widgets[$id] = $Widget;
			}
		} else {
				throw new WookieConnectorException('Problem getting available widgets');
			}
			
	 } catch(WookieConnectorException $e) {
			echo $e->errorMessage();
		}
    return $widgets;
  }
    
}
?>

File: connector/python/wookie/WookieConnectorService.py
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#
from xml.dom import minidom
import httplib
import urllib
httplib.HTTPConnection.debugLevel = 1
from WookieServerConnection import WookieServerConnection
from wookie.widget import Widget
from wookie.widget import Instance
from wookie.widget import Instances
from wookie.widget import User
from wookie.widget import Property


class WookieConnectorService:
    __connection = ""
    __widgetInstances = Instances.Instances()
    __currentUser = ""

    def __init__(self, wookieUrl, wookiePath = "", api_key = "", shareddatakey = ""):
        self.__connection = WookieServerConnection(wookieUrl, api_key, shareddatakey, wookiePath)

    # set current user
    # @param username, screenname

    def setUser(self, loginName, screenName = ""):
        if screenName == "":
            screenName = loginName
        self.__currentUser = User.User(loginName, screenName)

    # get current use
    # @return User object

    def getCurrentUser(self):
        return self.__currentUser

    # get current connection object
    # @return WookieServerConnection
    
    def getConnection(self):
        return self.__connection
        
    # get all available widgets
    # @return list of widgets
    
    def getAvailableWidgets(self):
        socket = httplib.HTTPConnection(self.getConnection().getUrl())
        socket.request('GET', self.getConnection().getPath()+'/widgets?all=true')
        response = socket.getresponse()
        xmldoc = ''
        try:
            xmldoc = minidom.parseString(response.read());
        except Exception:
            print 'Error getting widgets XML'
        socket.close()
        
        ##define widgetList
        widgetList = []
        if xmldoc:
            widgetsAvailable = xmldoc.getElementsByTagName('widget')
            if widgetsAvailable:
                for widget in widgetsAvailable:
                    
                    ##get widget title
                    widgetTitle = self.getText(widget.getElementsByTagName('title'));
                        
                     ##get widget url
                    widgetGuid = widget.getAttribute('identifier')
                        
                    ##get widget description
                    widgetDesc = self.getText(widget.getElementsByTagName('description'));
                        
                    ##get widget icon
                    widgetIcon = self.getText(widget.getElementsByTagName('icon'));
                    
                    if widgetTitle and widgetGuid:
                      widgetList.append(Widget.Widget(widgetGuid, widgetTitle, widgetDesc, widgetIcon))
            else:
                print 'No widgets found'
        return widgetList

    # Function to get Text value from XML node
    # @param parentElement
    # @return string

    def getText(self, parentElement):
        try:
            node = parentElement[0].childNodes[0]
            if node.nodeType == node.TEXT_NODE:
                return node.nodeValue
        except IndexError:
            pass
        return ''

    # Get or create widget instance
    # @param Widget instance or plain text GUID
    # @return new widget instance

    def getOrCreateInstance(self, widget_or_guid):
        if widget_or_guid != str(widget_or_guid):
            guid = widget_or_guid.getGuid()
        else:
            guid = widget_or_guid
        if self.getCurrentUser().getLoginName() == "":
            print "Current user loginName value empty (\"\")"
            return
        if guid:
            params = urllib.urlencode({'api_key': self.getConnection().getApiKey(),
                                       'userid': self.getCurrentUser().getLoginName(),
                                       'shareddatakey': self.getConnection().getSharedDataKey(),
                                       'widgetid': guid})
            headers = {"Content-type": "application/x-www-form-urlencoded",
                       "Accept": "text/xml"}
            socket = httplib.HTTPConnection(self.getConnection().getUrl())
            socket.request('POST', self.getConnection().getPath()+'/widgetinstances', params, headers)
            try:
                response = socket.getresponse()
            except ResponseNotReady:
                response = socket.getresponse()
                
            if response.status == 201:
                response = socket.getresponse()
            instanceXml = response.read()
            socket.close()
            newInstance = self.parseInstance(instanceXml, guid)
            if newInstance:
                self.__widgetInstances.put(newInstance)
                self.addParticipant(newInstance, self.getCurrentUser())
                return newInstance
            else:
                return

    # Parse widget instance XML
    # @param xml, guid
    # @return new widget instance
        
    def parseInstance(self, xml, guid):
        newInstance = ''
        xmlDoc = ''
        try:
            xmlDoc = minidom.parseString(xml)
        except Exception:
            print 'Could not parse instance xml'
        if xmlDoc:
            url = self.getText(xmlDoc.getElementsByTagName('url'))
            title = self.getText(xmlDoc.getElementsByTagName('title'))
            height = self.getText(xmlDoc.getElementsByTagName('height'))
            width = self.getText(xmlDoc.getElementsByTagName('width'))
            isMaximizable = self.getText(xmlDoc.getElementsByTagName('maximize'))
            newInstance = Instance.Instance(url, guid, title, height, width, isMaximizable)
        return newInstance

    # Get list of participants
    # @param widget instance
    # @return List participants

    def getParticipants(self, widgetInstance):
        participantsList = []
        if widgetInstance == "":
            print "No widget instance"
            return
        queryString = '?api_key='+self.getConnection().getApiKey()
        queryString += '&userid='+self.getCurrentUser().getLoginName()
        queryString += '&shareddatakey='+self.getConnection().getSharedDataKey()
        queryString += '&widgetid='+widgetInstance.getGuid()
        
        socket = httplib.HTTPConnection(self.getConnection().getUrl())
        socket.request('GET', self.getConnection().getPath()+'/participants'+queryString)
        try:
            response = socket.getresponse()
        except ResponseNotReady:
            response = socket.getresponse()

        if response.status == 200:
            xmlDoc = ""
            try:
                xmlDoc = minidom.parseString(response.read())
                participants = xmlDoc.getElementsByTagName("participant")
                if participants:
                    for participant in participants:
                        participant_id = participant.getAttribute("id")
                        display_name = participant.getAttribute("display_name")
                        thumbnailURL = participant.getAttribute("thumbnail_url")
                        newUser = User.User(participant_id, display_name)
                        newUser.setThumbnail(thumbnailURL)
                        participantsList.append(newUser)
            except Exception:
                print "Could not parse participants XML"
            
            return participantsList
        if response.status != 200:
            print "HTTP Status: "+response.status+"\nResponseText: "+response.read()
            return
        socket.close()
        return participantsList

    # Add participant to current widget instance
    # @param widgetInstance, userInstance
    # @return true if added, false otherwise
    
    def addParticipant(self, widgetInstance, userInstance):
        params = urllib.urlencode({'api_key': self.getConnection().getApiKey(),
                                   'userid': self.getCurrentUser().getLoginName(),
                                   'shareddatakey': self.getConnection().getSharedDataKey(),
                                   'widgetid': widgetInstance.getGuid(),
                                   'participant_id': userInstance.getLoginName(),
                                   'participant_display_name': userInstance.getScreenName(),
                                   'participant_thumbnail_url': userInstance.getThumbnail()})
        
        headers = {"Content-type": "application/x-www-form-urlencoded",
                       "Accept": "text/xml"}
        socket = httplib.HTTPConnection(self.getConnection().getUrl())
        socket.request('POST', self.getConnection().getPath()+'/participants', params, headers)
        try:
            response = socket.getresponse()
        except ResponseNotReady:
            response = socket.getresponse()
        if response.status == 201:
            return 'true'
        if response.status == 200:
            return 'true'
        return 'false'

    # Delete participant
    # @param widgetIntance, userInstance
    # @return true if done, false if failed

    def deleteParticipant(self, widgetInstance, userInstance):
        if widgetInstance == "":
            print "No widget instance"
            return
        if userInstance == "":
            print "No user instance"
            return
        queryString = '?api_key='+self.getConnection().getApiKey()
        queryString += '&userid='+self.getCurrentUser().getLoginName()
        queryString += '&shareddatakey='+self.getConnection().getSharedDataKey()
        queryString += '&widgetid='+widgetInstance.getGuid()
        queryString += '&participant_id='+userInstance.getLoginName()
        
        socket = httplib.HTTPConnection(self.getConnection().getUrl())
        socket.request('DELETE', self.getConnection().getPath()+'/participants'+queryString)
        try:
            response = socket.getresponse()
        except ResponseNotReady:
            response = socket.getresponse()

        if response.status == 200:
            return 'true'
        if response.status == 404:
            return 'false'

        print "deleteParticipant error: \n"+response.status
        return

    # Set widget property
    # @param widgetInstance, property instance
    # @return true or false

    def setProperty(self, widgetInstance, propertyInstance):
        if widgetInstance == "":
            print "No widget instance"
            return
        if propertyInstance == "":
            print "No property instance"
            return
        params = urllib.urlencode({'api_key': self.getConnection().getApiKey(),
                                   'userid': self.getCurrentUser().getLoginName(),
                                   'shareddatakey': self.getConnection().getSharedDataKey(),
                                   'widgetid': widgetInstance.getGuid(),
                                   'propertyname': propertyInstance.getName(),
                                   'propertyvalue': propertyInstance.getValue(),
                                   'is_public': propertyInstance.getIsPublic()})
        
        headers = {"Content-type": "application/x-www-form-urlencoded",
                       "Accept": "text/plain"}
        socket = httplib.HTTPConnection(self.getConnection().getUrl())
        socket.request('POST', self.getConnection().getPath()+'/properties', params, headers)
        try:
            response = socket.getresponse()
        except ResponseNotReady:
            response = socket.getresponse()
        if response.status == 200:
            return 'true'
        if response.status == 201:
            return 'true'
        if response.status > 201:
            print "setProperty error: "+response.status
        return 'false'

    # Delete property
    # @param widgetInstance, property instance
    # @return true or false

    def deleteProperty(self, widgetInstance, propertyInstance):
        if widgetInstance == "":
            print "No widget instance"
            return
        if propertyInstance == "":
            print "No property instance"
            return
        queryString = '?api_key='+self.getConnection().getApiKey()
        queryString += '&userid='+self.getCurrentUser().getLoginName()
        queryString += '&shareddatakey='+self.getConnection().getSharedDataKey()
        queryString += '&widgetid='+widgetInstance.getGuid()
        queryString += '&propertyname='+propertyInstance.getName()
        
        socket = httplib.HTTPConnection(self.getConnection().getUrl())
        socket.request('DELETE', self.getConnection().getPath()+'/properties'+queryString)
        try:
            response = socket.getresponse()
        except ResponseNotReady:
            response = socket.getresponse()
        if response.status == 200:
            return 'true'
        if response.status == 404:
            return 'false'
        print "deleteProperty error: "+response.status
        return 'false'

    # Get property
    # @param widget instance, property instance
    # @return retrieved property instance

    def getProperty(self, widgetInstance, propertyInstance):
        if widgetInstance == "":
            print "No widget instance"
            return
        if propertyInstance == "":
            print "No property instance"
            return
        queryString = '?api_key='+self.getConnection().getApiKey()
        queryString += '&userid='+self.getCurrentUser().getLoginName()
        queryString += '&shareddatakey='+self.getConnection().getSharedDataKey()
        queryString += '&widgetid='+widgetInstance.getGuid()
        queryString += '&propertyname='+propertyInstance.getName()
        
        socket = httplib.HTTPConnection(self.getConnection().getUrl())
        socket.request('GET', self.getConnection().getPath()+'/properties'+queryString)
        try:
            response = socket.getresponse()
        except ResponseNotReady:
            response = socket.getresponse()
        propertyValue = response.read()
        if response.status == 200:
            return Property.Property(propertyInstance.getName(), propertyValue)
        print "getProperty error: "+response.status
        return 'false'
